cmd/expiration-mailer/main.go

```diff
@@ -205,88 +212,115 @@ func (m *mailer) processCerts(ctx context.Context, allCerts []core.Certificate)
-			if err != nil {
-				// TODO(#1420): tell registration about this error
-				m.log.AuditErrf("Error parsing certificate %s: %s", cert.Serial, err)
-				m.stats.errorCount.With(prometheus.Labels{"type": "ParseCertificate"}).Inc()
-				continue
-			}
-
-			renewed, err := m.certIsRenewed(ctx, parsedCert.DNSNames, parsedCert.NotBefore)
-			if err != nil {
-				m.log.AuditErrf("expiration-mailer: error fetching renewal state: %v", err)
-				// assume not renewed
-			} else if renewed {
-				m.log.Debugf("Cert %s is already renewed", cert.Serial)
-				m.stats.certificatesAlreadyRenewed.Add(1)
-				err := m.updateCertStatus(ctx, cert.Serial)
+		wg.Add(1)
+		go func(conn bmail.Conn, ch <-chan work) {
+			defer wg.Done()
+			for w := range ch {
+				err := m.sendToOneRegID(ctx, conn, w.regID, w.certs)
 				if err != nil {
-					m.log.AuditErrf("Error updating certificate status for %s: %s", cert.Serial, err)
-					m.stats.errorCount.With(prometheus.Labels{"type": "UpdateCertificateStatus"}).Inc()
+					m.log.AuditErr(err.Error())
 				}
-				continue
 			}
+			conn.Close()
+		}(conn, workChan)
+	}
+	for regID, certs := range regIDToCerts {
+		workChan <- work{regID, certs}
+	}
+	close(workChan)
+	wg.Wait()
+	return nil
+}
 
-			parsedCerts = append(parsedCerts, parsedCert)
-		}
+func (m *mailer) sendToOneRegID(ctx context.Context, conn bmail.Conn, regID int64, certs []core.Certificate) error {
+	reg, err := m.rs.GetRegistration(ctx, &sapb.RegistrationID{Id: regID})
+	if err != nil {
+		m.stats.errorCount.With(prometheus.Labels{"type": "GetRegistration"}).Inc()
+		return fmt.Errorf("fetching registration %d: %w", regID, err)
+	}
 
-		m.stats.certificatesPerAccountNeedingMail.Observe(float64(len(parsedCerts)))
+	if reg.Contact == nil {
+		return nil
+	}
 
-		if len(parsedCerts) == 0 {
-			// all certificates are renewed
-			continue
+	parsedCerts := []*x509.Certificate{}
+	for _, cert := range certs {
+		parsedCert, err := x509.ParseCertificate(cert.DER)
+		if err != nil {
+			m.stats.errorCount.With(prometheus.Labels{"type": "ParseCertificate"}).Inc()
+			// TODO(#1420): tell registration about this error
+			return fmt.Errorf("parsing certificate %s: %w", cert.Serial, err)
 		}
 
-		// TODO(#6121): Remove this
-		if !features.Enabled(features.ExpirationMailerDontLookTwice) {
-			if len(reg.Contact) == 0 {
-				continue
+		renewed, err := m.certIsRenewed(ctx, parsedCert.DNSNames, parsedCert.NotBefore)
+		if err != nil {
+			return fmt.Errorf("expiration-mailer: error fetching renewal state: %w", err)
+		} else if renewed {
+			m.stats.certificatesAlreadyRenewed.Add(1)
+			err := m.updateCertStatus(ctx, cert.Serial)
+			if err != nil {
+				m.stats.errorCount.With(prometheus.Labels{"type": "UpdateCertificateStatus"}).Inc()
+				return fmt.Errorf("updating certificate status for %s: %w", cert.Serial, err)
 			}
+			continue
 		}
 
-		err = m.sendNags(reg.Contact, parsedCerts)
+		parsedCerts = append(parsedCerts, parsedCert)
+	}
+
```

Comment:
We lost `		m.stats.certificatesPerAccountNeedingMail.Observe(float64(len(parsedCerts)))` here, effectively reverting #6091.